File: JavaSwiftKitDemo/src/main/java/com/example/swift/HelloSwift.java
//===----------------------------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2024 Apple Inc. and the Swift.org project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
// See CONTRIBUTORS.txt for the list of Swift.org project authors
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

package com.example.swift;

import com.example.swift.HelloSubclass;

public class HelloSwift {
    private double value;
    private static double initialValue = 3.14159;
    private String name = "Java";

    static {
        System.loadLibrary("JavaKitExample");
    }

    public HelloSwift() {
        this.value = initialValue;
    }

    public static void main(String[] args) {
        int result = new HelloSubclass("Swift").sayHello(17, 25);
        System.out.println("sayHello(17, 25) = " + result);
    }

    public native int sayHello(int x, int y);
    public native String throwMessageFromSwift(String message) throws Exception;

    // To be called back by the native code
    private double sayHelloBack(int i) {
        System.out.println("And hello back from " + name + "! You passed me " + i);
        return value;
    }

    public void greet(String name) {
        System.out.println("Salutations, " + name);
    }

    String[] doublesToStrings(double[] doubles) {
        int size = doubles.length;
        String[] strings = new String[size];

        for(int i = 0; i < size; i++) {
            strings[i] = "" + doubles[i];
        }

        return strings;
    }

    public void throwMessage(String message) throws Exception {
        throw new Exception(message);
    }
}

File: Samples/JavaKitSampleApp/Sources/JavaKitAndroidExample/app/src/main/java/com/example/swift/HelloSwift.kt
package com.example.swift

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.enableEdgeToEdge
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import com.example.swift.ui.theme.JavaKitSampleAppTheme
import java.util.function.Predicate

open class HelloSwift : ComponentActivity() {
    init {
        System.loadLibrary("JavaKitExample")
    }

    external fun sayHello(x: Int, y: Int): Int
    external fun throwMessageFromSwift(message: String)

    private fun getGreeting(): String {
        val result = sayHello(17, 25)
        return "$result"
    }

    var value: Double = 0.0
    var name: String = "Java"

    fun sayHelloBack(i: Int): Double {
        return value
    }

    fun greet(name: String) {
        println("Salutations, $name")
    }

    fun lessThanTen(): Predicate<Int> {
        val predicate = Predicate<Int> { it < 10 }
        return predicate
    }

    fun doublesToStrings(doubles: DoubleArray): Array<String?> {
        val size = doubles.count()
        var strings: Array<String?> = arrayOfNulls(doubles.count())

        doubles.forEachIndexed { index, element ->
            strings[index] = "$element"
        }

        return strings
    }

    fun throwMessage(message: String) {
        throw Exception(message)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContent {
            JavaKitSampleAppTheme {
                Scaffold(modifier = Modifier.fillMaxSize()) { innerPadding ->
                    Greeting(
                        name = getGreeting(),
                        modifier = Modifier.padding(innerPadding)
                    )
                }
            }
        }
    }
}

@Composable
fun Greeting(name: String, modifier: Modifier = Modifier) {
    Text(
        text = "sayHello(17, 25) = $name",
        modifier = modifier
    )
}

@Preview(showBackground = true)
@Composable
fun GreetingPreview() {
    JavaKitSampleAppTheme {
        Greeting("Android")
    }
}

class HelloSubclass(private val greeting: String = "Swift"): HelloSwift() {
    fun greetMe() {
       super.greet(greeting)
    }
}